### ISSUE TITLE
Editor Themes: Product page does not display Klarna options

### ISSUE THREAD

**Issue by sophiegyo** [Automattic/themes#7526]
### Quick summary

Summary: When the Tsubaki theme is active on a site, Klarna options are not displayed on the product page.
This does not occur with non-Site-Editor themes that don't use blocks for the product page template.

User report in: 7400853-zen

I'm guessing this is related to the [cart and checkout page issue](https://github.com/Automattic/wp-calypso/issues/80926) for Tsubaki.

### Steps to reproduce

1. Head to the user site (noted in 7400853-zen) as a regular visitor, while Tsubaki is active.
2. Check any product.
3. Klarna options are not shown on the product page (but they are currently showing at checkout)

This can be consistently reproduced.
I tested with the Health Check & Troubleshooting plugin to deactivate everything else.
In Troubleshooting Mode, with ONLY Woo, Klarna, and Tsubaki active, the Klarna options do not show.

### What you expected to happen

The Klarna financing option should be shown on the product page.

For comparison, how it shows up when using Storefront. Specifically: Troubleshooting mode, with only Woo, Storefront, and Klarna active:
![image](https://github.com/Automattic/themes/assets/36608681/964c6c22-87f7-4e13-a027-14e1547292ef)

And for good measure: Troubleshooting mode with only Woo, 3pd theme Aotwo, and Klarna active:
![image](https://github.com/Automattic/themes/assets/36608681/cae0bb94-1659-4e27-a3aa-836a4ab18644)

### What actually happened

It doesn't appear. Screenshot:
![image](https://github.com/Automattic/themes/assets/36608681/ad90b6ec-e32f-479c-9972-c0ff6504f145)

### Browser

Google Chrome/Chromium, Mozilla Firefox

### Context

Customer report 7400853-zen

### Platform (Simple, Atomic, or both?)

Atomic

### Other notes

I haven't been able to set up test Klarna credentials of my own to try to reproduce this on a test site, but I'm pretty confident it's the same as the [issue](https://github.com/Automattic/wp-calypso/issues/80926) mentioned earlier.

### Reproducibility

Consistent

### Severity

Some (< 50%)

### Available workarounds?

Yes, easy to implement

### Workaround details

The only workaround that I can see would be to switch to a non-Site-Editor theme - or at least any theme which uses a classic PHP template for the product page.

**Comment by nerrad** [comment]
It'd be good to see if this can be reproduced with other block themes. 
@PanosSynetos, my guess is there could be an issue with the single product page block template compatibility if this is also reproducible in other Block themes with no customized Woo Single Product template. If that's the case we may need to work with whatever team helps oversee the Klarna extension to figure out what is happening (what hook is being used etc).

**Comment by nerrad** [comment]
@danielwrobert please have someone from Kirigami or Origami do some preliminary investigation on this as well. If this is a block template compatibility issue there will be other things that will be affected besides Klarna.

**Comment by gigitux** [comment]
I'm able to replicate the issue. You have to:

- [Klarna Payments for WooCommerce](https://it.wordpress.org/plugins/klarna-payments-for-woocommerce/).
- Configure it.
- Install [Klarna On Site messaging](https://docs.krokedil.com/klarna-for-woocommerce/additional-klarna-plugins/klarna-on-site-messaging/).
- Configure it.


With Storefront, the placeholder is visible. On TT4, it isn't visible.


| Storefront | TT4 |
|--------|--------|
|![image](https://github.com/Automattic/themes/assets/4463174/d89ccb2d-9217-46a2-ae62-2b80cfb793e0)| ![image](https://github.com/Automattic/themes/assets/4463174/cd307d7f-e09b-4ee0-a3b3-7c570e0a5bbb) | 


Checking `Klarna On Site messaging` codebase,  I found the issue. The class that injects the Klarna content is initialized during the `wp_head` hook.  ([source code](https://github.com/krokedil/klarna-onsite-messaging-for-woocommerce/blob/d9562ad79c64b5e5e8383a3085769937f7316ca7/classes/class-klarna-onsite-messaging-product-page.php/#L32-L33)). When the `wp_head` is run, the logic of the compatibility layer has already been executed. The compatibility layer logic runs when WordPress retrieves the right template to render. [https://github.com/woocommerce/woocommerce/blob/76d43014a373a79a888bff470bd5c24aae[…]2c/plugins/woocommerce/src/Blocks/BlockTemplatesController.php/](https://github.com/woocommerce/woocommerce/blob/76d43014a373a79a888bff470bd5c24aae07832c/plugins/woocommerce/src/Blocks/BlockTemplatesController.php/#L62-L63)
The action `wp_head` is too late because WordPress already retrieved the template. 

If I updated to:

```php
add_action( 'wp', array( $this, 'init_class' ) );
``` 

The Klarna placeholder is injected correctly on a block theme as well:

![image](https://github.com/Automattic/themes/assets/4463174/937db8cd-dac0-45b1-86ec-480aa271c381)



Furthermore, I noticed that the plugin proposes different positions in the settings, but with block themes, the setting will not work. The plugin logic is based on priority, but this approach doesn't work with the compatibility layer. ([source code](https://github.com/krokedil/klarna-onsite-messaging-for-woocommerce/blob/d9562ad79c64b5e5e8383a3085769937f7316ca7/classes/class-klarna-onsite-messaging-product-page.php/#L48-L49))

**Comment by Robertght** [comment]
I tested with TT4 and this is what I saw:

![Image](https://github.com/user-attachments/assets/ecc41536-117a-4da3-ba6c-6d189dd0943e)

And with Tsubaki:

![Image](https://github.com/user-attachments/assets/663fc593-117c-4e03-b10e-1aa19462f7ec)

Some things to note:

- [Klarna On Site messaging](https://docs.krokedil.com/klarna-for-woocommerce/additional-klarna-plugins/klarna-on-site-messaging/) is retired so it's already on on my site
- I tested this using a playground/test environment
- I can't see the financing model even with Storefront on the product page
- I can see the express checkout but not the other options

I'm not sure what's actually wrong with my setup, however, @sophiegyo or @gigitux do you have an environment to see if this is still happening on your end please?

**Comment by gigitux** [comment]
@Robertght I don't think that you see the `Klarna On Site messaging`, but you see the `Express Checkout`:

![Image](https://github.com/user-attachments/assets/bafeb9a7-6421-464b-a3ce-cd72f0e220aa)

Could you try to enable a classic theme and upload a screenshot of a product page? Thanks!

cc @imanish003 @WunderBart

**Comment by Robertght** [comment]
>I don't think that you see the Klarna On Site messaging, but you see the Express Checkout:

Yes, as I mentioned above, that's what I'm seeing.

>Could you try to enable a classic theme and upload a screenshot of a product page? Thanks!

The above example includes storefront too, however, it feels like there's an issue with my actual setup despite everything in the Klarna dashboard showing as "active". Is this perhaps a geo-related thing or should I just create a new site for it?

Test product: https://atomicbobv22018.wpcomstaging.com/product/booking-product/beanie-with-logo/

**Comment by Robertght** [comment]
After thorough troubleshooting around my test environment, it looks like the issue remains relevant.

Testing with TwentyTwentyFour does not display the Klarna banner, but at least the Express checkout button shows up.

### Storefront:

![Image](https://github.com/user-attachments/assets/310af027-33e1-40bb-a0c3-9281fce2c19d)

### TT4:

![Image](https://github.com/user-attachments/assets/5acfd6d9-0d81-4929-ae67-9f9c045af5a5)

props to @gigitux for assisting me in setting this up.
